File: fm-cli/go-projector/pkg/projector/projector_test.go
package projector_test

import (
	"testing"

	"example.com/go-projector/pkg/projector"
)

func getData() *projector.Data {
	return &projector.Data{
		Projector: map[string]map[string]string{
			"/": {
				"foo": "bar1",
			},
			"/foo": {
				"foo": "bar2",
			},
			"/foo/bar": {
				"foo": "bar3",
				"bar": "foo3",
			},
		},
	}
}

func getProjector(pwd string, data *projector.Data) *projector.Projector {
	return projector.CreateProjector(&projector.Config{
		Args:      []string{},
		Pwd:       pwd,
		Config:    "HELLO WORLD",
		Operation: projector.Print,
	}, data)
}

func test(t *testing.T, proj *projector.Projector, key, value string) {

	v, ok := proj.GetValue(key)

	if !ok {
		t.Errorf("expected to find value %q", value)
	}

	if value != v {
		t.Errorf("Expected value to be %v, got %v", value, v)
	}

}

func TestGetValue(t *testing.T) {
	data := getData()
	p := getProjector("/foo/bar", data)
	test(t, p, "bar", "foo3")
}

func TestSetValue(t *testing.T) {
	data := getData()
	p := getProjector("/foo/bar", data)
	test(t, p, "bar", "foo3")
	p.SetValue("bar", "foo4")
	test(t, p, "bar", "foo4")
	p.SetValue("fm", "foo4")
	p = getProjector("/", data)
	test(t, p, "foo", "bar1")
}

func TestRemoveValue(t *testing.T) {
	data := getData()
	p := getProjector("/foo/bar", data)
	test(t, p, "foo", "bar3")
	p.RemoveValue("foo")
	test(t, p, "foo", "bar2")
}

File: projector/ts-projector/src/projector.test.ts
import { Operation } from './config';
import Projector, { Data } from './projector';

function getData(): Data {
  return {
    "projector": {
      "/": {
        "foo": "bar1",
      },
      "/foo": {
        "foo": "bar2",
      },
      "/foo/bar": {
        "foo": "bar3",
        "bar": "foo3",
      },
    }
  }
}

function getProjector(pwd: string, data = getData()): Projector {
  return new Projector({
    pwd,
    args: [],
    operation: Operation.Print,
    config: "HELLO WORLD",
  }, data);
}



it("getValueAll", () => {
  const projector = getProjector("/foo/bar");
  expect(projector.getValueAll()).toEqual({
    "foo": "bar3",
    "bar": "foo3",
  });
})

it("getValue", () => {
  let projector = getProjector("/foo/bar");
  expect(projector.getValue("foo")).toEqual("bar3");
  expect(projector.getValue("bar")).toEqual("foo3");

  projector = getProjector("/");
  expect(projector.getValue("foo")).toEqual("bar1");

  projector = getProjector("/foo");
  expect(projector.getValue("foo")).toEqual("bar2");
})

it("setValue", () => {
  let data = getData();
  let projector = getProjector("/foo/bar", data);
  projector.setValue("foo", "baz");

  expect(projector.getValue("foo")).toEqual("baz");
  projector.setValue("foo4", "baz4");
  expect(projector.getValue("foo4")).toEqual("baz4");

  projector = getProjector("/", data);
  expect(projector.getValue("foo")).toEqual("bar1");
})

it("removeValue", () => {
  const projector = getProjector("/foo/bar", getData());
  projector.removeValue("foo");
  expect(projector.getValue("foo")).toEqual("bar2");
  projector.removeValue("bar");
  expect(projector.getValue("bar")).toEqual("");
})
